### DISCUSSION TITLE
Sharing a fuzzer for Burn

### DISCUSSION THREAD

**Discussion by SRaswan** [tracel-ai/burn#5660] [Show and tell]
Hi! I wanted to just share a [differential fuzzer](https://github.com/SRaswan/burnmark/) for Burn (for both tensor operations and automatic differentiation) that I think the team might find interesting. This was tested against 0.20.1 a while ago and uncovered a few bugs. It randomly generates ASTs for Burn to compile and test against Burn's LibTorch backend.

https://github.com/SRaswan/burnmark

Here is a [presentation](https://docs.google.com/presentation/d/1WGRMCECDWtMQh4U7GBAEMhACA0tHrbwKk16jniXJrhI/edit?usp=sharing) we made about it.

One of the bugs it uncovered, for example, was a memory management bug within Burn’s LibTorch backend binding in 0.20.1 (that is now fixed in 0.22!).

The randomly generated function looked like:
```rust
fn transpose_repro_loss<B: AutodiffBackend >(logits_flat : Tensor <B, 2>) -> Tensor <B, 1> {
  let t0 = logits_flat.clone().transpose().matmul(logits_flat);
  let t1 = t0.clone().transpose();
  let t2 = t1.clone() + t0.clone();
  (t2.clone() * t2).mean()
}
```

And it reproduced:
```
===========================================================
Model : TINY - TRANSPOSE
=== PANIC ===
A fatal error happened, thread ’main ’ (71928) panicked at .../ burn-tch-0.20.1/src/ops/base.rs:361:40:
called ’Result::unwrap()’ on an ’Err’ value: Torch(" unsupported operation: some elements of the input tensor and the written-to tensor refer to a single memory location. Please clone() the tensor before performing the operation. \ nException raised from assert_no_partial_overlap ... "
```

The bug stems from how the FFI handles tensor cloning and memory views. In tensor frameworks, operations that modify a tensor’s view (such as a transpose) require a proper clone if the tensor is being modified in-place without altering the original underlying data.

Through our fuzzer’s automated forward and backward pass evaluations, we discovered that Burn’s LibTorch binding was incorrectly performing a shallow clone rather than a deep copy, despite the explicit clone() calls. When the transposed view was modified or used in subsequent graph operations, the shallow clone caused unintended memory sharing across the FFI boundary, leading to incorrect gradient calculations and memory corruption.

The resolution to this bug is what is in the code now. The backend can now safely detect when memory sharing exists and force a deep copy during in-place view modifications.

```rust
// Before ( Buggy ): The new tensor loses its connection to the original storage
pub fn swap_dims(tensor: TchTensor, dim1: usize, dim2: usize) -> TchTensor {
  let tensor = tensor.tensor.transpose(dim1 as i64, dim2 as i64 );
  TchTensor::new(tensor)
}

// Now ( Fixed ): The new tensor explicitly shares the parent’s storage
pub fn swap_dims (tensor: TchTensor, dim1: usize, dim2: usize ) -> TchTensor {
  let storage = tensor.storage.clone();
  let tensor = tensor.tensor.transpose(dim1 as i64, dim2 as i64);
  TchTensor::from_existing(tensor, storage)
}
```

Just wanted to share this! I hope to also extend in the future with more rigorous testing against 0.22 and with more features (like higher-rank operations, more tensor operations, type-directed generation, etc.).

**Comment by laggui** [comment]
Thanks for sharing! This seems particularly useful for catching interactions between views, autodiff and more that isolated operator tests can miss.

It could also help validate backend contracts, including consistent edge-case behavior such as NaN propagation.

I could see something like this running periodically, maybe weekly with a time limit, against current Burn and available backends. Reproducible failures could then be minimized and added to our regression tests, broadening coverage without running the full fuzzer on every PR.
